xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/email/live/LiveMimeMessageIterator.java

```diff
@@ -0,0 +1,139 @@
+import org.xwiki.notifications.NotificationFormat;
+import org.xwiki.notifications.NotificationPreference;
+import org.xwiki.notifications.internal.ModelBridge;
+import org.xwiki.notifications.internal.NotificationFilterManager;
+import org.xwiki.notifications.internal.email.AbstractMimeMessageIterator;
+import org.xwiki.notifications.internal.email.NotificationUserIterator;
+
+/**
+ * MimeMessageIterator for sending live mail notifications.
+ *
+ * @since 9.6RC1
+ * @version $Id$
+ */
+@Component(roles = LiveMimeMessageIterator.class)
+@InstantiationStrategy(ComponentInstantiationStrategy.PER_LOOKUP)
+public class LiveMimeMessageIterator extends AbstractMimeMessageIterator
+{
+    private CompositeEvent compositeEvent;
+
+    @Inject
+    private NotificationFilterManager notificationFilterManager;
+
+    @Inject
+    @Named("cached")
+    private ModelBridge modelBridge;
+
+    @Inject
+    private Logger logger;
+
+    /**
+     * Initialize the iterator.
+     *
+     * @param userIterator iterator that returns all users
+     * @param factoryParameters parameters for the email factory
+     * @param event the event that has to be sent to every user
+     * @param templateReference reference to the mail template
+     */
+    public void initialize(NotificationUserIterator userIterator, Map<String, Object> factoryParameters,
+            CompositeEvent event, DocumentReference templateReference)
+    {
+        this.compositeEvent = event;
+        super.initialize(userIterator, factoryParameters, templateReference);
+    }
+
+    /**
+     * For the given user, we will have to check that the composite event that we have to send matches the user
+     * preferences.
+     *
+     * If, for any reason, one of the events of the original composite event is not meant for the user, we clone
+     * the original composite event and remove the incriminated event.
+     */
+    protected List<CompositeEvent> retrieveCompositeEventList(DocumentReference user) throws NotificationException
+    {
+        CompositeEvent resultCompositeEvent = new CompositeEvent(this.compositeEvent);
+
+        if (this.hasCorrespondingNotificationPreference(user, resultCompositeEvent)) {
+            // Apply the filters that the user has defined in its notification preferences
+            // If one of the events present in the composite event does not match a user filter, remove the event
+            for (NotificationFilter filter : this.notificationFilterManager.getAllNotificationFilters(user)) {
+                Iterator<Event> it = resultCompositeEvent.getEvents().iterator();
+                while (it.hasNext()) {
+                    Event event = it.next();
+                    if (filter.filterEvent(event, user, NotificationFormat.EMAIL)) {
+                        it.remove();
+                        if (resultCompositeEvent.getEvents().size() == 0) {
+                            return Collections.emptyList();
```

Comment:
Well, why don't you return `resultCompositeEvent.getEvents()` anyway? Why don't you return `resultCompositeEvent` which has also the list?